doc/devs.rst

```diff
@@ -0,0 +1,321 @@
+default. Further, IP multicast is not routed in the Internet. Extensions for
+reliability and security exist, but we cannot rely on Broker users to provide
+this infrastructure. To run "out of the box", we rely on TCP connections only
+and minimize message duplication via ALM.
+
+Goals
+  - Connections are secured/authenticated through TLS & certificates.
+  - Broker puts no constraints on topologies, i.e., users may establish all
+    peering relations that make sense in their deployment without having to
+    form a tree. Loops are explicitly allowed.
+  - Broker detects (and slows down) any publisher that produces data at a faster
+    rate than the subscribers or the network can handle.
+  - Each peer can publish to each topic. No central coordinator or root node
+    exists.
+  - Broker minimizes messages in the network with (application-layer) multicast
+    in order to scale effectively.
+
+Non-Goals
+  - Dynamic connection management and auto-balancing of a distribution tree.
+    While Broker generally follows a P2P philosophy, we still wish to give users
+    full control over Broker's topology. Also, Broker does not target
+    Internet-wide deployments with very high churn rates and unreliable nodes.
+
+Peers
+~~~~~
+
+Each Broker peer in the network has:
+
+- A globally unique ID. Currently, we use ``caf::node_id`` to identify peers.
+  CAF computes this ID automatically by combining a 160-bit hash value (based on
+  a seed plus various node-specific information) with the OS-specific process
+  ID.
+- Local subscriptions. Endpoints combine the subscriptions of all local
+  subscribers to a single subscription list. The endpoint removes all redundant
+  entries from this list. For example, an endpoint with local subscribers for
+  ``/zeek/event/foo``, ``/zeek/event/bar``, and ``/zeek/event`` is only going to
+  store ``/zeek/event``. This subscription automatically includes the other two
+  due to the prefix matching.
+- A logical clock (Lamport timestamps). This 64-bit integer enables detection of
+  repeated messages and ordering of events. Whenever a peer sends a message to
+  others, it increments its logical time and includes the current value in the
+  message. This timestamp is crucial for detecting outdated or repeated
+  subscriptions in the `Subscription Flooding`_.
+- A routing table with paths to *all* known peers in the network as well as the
+  subscriptions of each peer.
+
+Subscription Flooding
+~~~~~~~~~~~~~~~~~~~~~
+
+Whenever the local subscriptions change, a peer sends a *subscription* message
+to all of its direct connections. When establishing a new peering relation, the
+handshake also includes the *subscription* message.
+
+The subscription message consists of:
+
+#. A ``peer_id_list`` for storing the path of this message. Initially, this list
+   only contains the ID of the sender.
+#. A ``filter_type`` (list of topics) that denotes the current subscriptions.
```

Comment:
The type is actually called `filter_type`: https://github.com/zeek/broker/blob/topic/neverlord/multi-hop-routing/include/broker/filter_type.hh. I think renaming it to `topic_list` would give a wrong intuition, since there's a bit more to the type. While the representation is "just" a vector at the moment, we might switch to a different data structure (such as radix trees) at some point. I'll update the docs to avoid calling it topic list at all.